homeassistant/components/yolink/__init__.py

```diff
@@ -49,12 +58,83 @@
 class YoLinkCoordinator(DataUpdateCoordinator):
     """YoLink DataUpdateCoordinator."""
 
+    yl_devices: list[YoLinkDevice] = []
+
     def __init__(
-        self,
-        hass: HomeAssistant,
+        self, hass: HomeAssistant, yl_client: YoLinkClient, yl_mqtt_client: MqttClient
     ) -> None:
-        """Init YoLink DataUpdateCoordinator."""
-        super().__init__(hass, _LOGGER, name=DOMAIN)
+        """Init YoLink DataUpdateCoordinator.
+
+        fetch state every 30 minutes base on yolink device heartbeat interval
+        """
+        super().__init__(
+            hass, _LOGGER, name=DOMAIN, update_interval=timedelta(minutes=30)
+        )
+        self._client = yl_client
+        self._mqtt_client = yl_mqtt_client
+        self.data = {}
+
+    async def init_coordinator(self):
+        """Init coordinator."""
+        try:
+            async with async_timeout.timeout(10):
+                home_info = await self._client.get_general_info()
+
+                def on_message_callback(message: tuple[str, BRDP]):
+                    """On message callback."""
+                    data = message[1]
+                    if data.event is None:
+                        return
+                    event_param = data.event.split(".")
+                    event_type = event_param[len(event_param) - 1]
+                    if event_type not in (
+                        "Report",
+                        "Alert",
+                        "StatusChange",
+                        "getState",
+                    ):
+                        return
+                    resolved_state = data.data
+                    if resolved_state is None:
+                        return
+                    self.data[message[0]] = resolved_state
+                    self.async_set_updated_data(self.data)
+
+                await self._mqtt_client.init_home_connection(
+                    home_info.data["id"], on_message_callback
+                )
+            async with async_timeout.timeout(10):
+                device_response = await self._client.get_auth_devices()
+                yl_devices: list[YoLinkDevice] = []
```

Comment:
Please only wrap the line that can raise within the try... except block. We can probably move these lines below the except blocks.